File: SamplePresence/UIKit-ObjC/SetupViewController.m
#import "SetupViewController.h"
#import "DemoFlowController.h"
#import "FormFieldCell.h"

typedef NS_ENUM(NSInteger, SetupSection) {
    SetupSectionAuthToken = 0,
    SetupSectionCreateSession = 1,
    SetupSectionRetrieveSession = 2,
    SetupSectionCount = 3
};

@interface SetupViewController ()
@property (nonatomic, weak) DemoFlowController *flowController;
@property (nonatomic, copy) NSString *pin;
@end

@implementation SetupViewController

- (instancetype)initWithFlowController:(DemoFlowController *)flowController {
    self = [super initWithStyle:UITableViewStyleInsetGrouped];
    if (self) {
        _flowController = flowController;
        _pin = @"";
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Setup Session";
    [self.tableView registerClass:[FormFieldCell class] forCellReuseIdentifier:FormFieldCell.reuseIdentifier];
    [self bindFlowController];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.flowController.errorMessage == nil ? SetupSectionCount : SetupSectionCount + 1;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    switch (section) {
        case SetupSectionAuthToken: return @"Auth Token";
        case SetupSectionCreateSession: return @"Create Session";
        case SetupSectionRetrieveSession: return @"Retrieve Session";
        default: return nil;
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (section == SetupSectionCount) {
        return 1;
    }
    switch (section) {
        case SetupSectionAuthToken: return 1;
        case SetupSectionCreateSession: return 2;
        case SetupSectionRetrieveSession: return 2;
        default: return 0;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == SetupSectionCount) {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ErrorCell"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"ErrorCell"];
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
            cell.textLabel.numberOfLines = 0;
        }
        cell.textLabel.text = self.flowController.errorMessage;
        cell.textLabel.textColor = UIColor.systemRedColor;
        return cell;
    }

    if (indexPath.section == SetupSectionAuthToken) {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TokenCell"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"TokenCell"];
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
            cell.textLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
            cell.textLabel.numberOfLines = 0;
        }
        cell.textLabel.text = self.flowController.session.authToken;
        return cell;
    }

    if (indexPath.section == SetupSectionCreateSession) {
        if (indexPath.row == 0) {
            FormFieldCell *cell = [tableView dequeueReusableCellWithIdentifier:FormFieldCell.reuseIdentifier forIndexPath:indexPath];
            [cell configureWithTitle:@"Contact Email:"
                                text:self.flowController.session.contactEmail
                         placeholder:@"contact@example.com"
                        keyboardType:UIKeyboardTypeEmailAddress];
            [cell.valueField addTarget:self action:@selector(contactEmailChanged:) forControlEvents:UIControlEventEditingChanged];
            return cell;
        }
        return [self actionCellWithTitle:self.flowController.isBusy ? @"Creating…" : @"Create Session"
                                 enabled:!self.flowController.isBusy];
    }

    if (indexPath.row == 0) {
        FormFieldCell *cell = [tableView dequeueReusableCellWithIdentifier:FormFieldCell.reuseIdentifier forIndexPath:indexPath];
        [cell configureWithTitle:@"PIN:" text:self.pin placeholder:@"Required" keyboardType:UIKeyboardTypeDefault];
        [cell.valueField addTarget:self action:@selector(pinChanged:) forControlEvents:UIControlEventEditingChanged];
        return cell;
    }

    BOOL enabled = !self.flowController.isBusy && self.pin.length > 0;
    return [self actionCellWithTitle:self.flowController.isBusy ? @"Retrieving…" : @"Retrieve Session"
                             enabled:enabled];
}

- (UITableViewCell *)actionCellWithTitle:(NSString *)title enabled:(BOOL)enabled {
    UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:nil];
    UIListContentConfiguration *configuration = [UIListContentConfiguration cellConfiguration];
    configuration.text = title;
    configuration.textProperties.alignment = NSTextAlignmentCenter;
    configuration.textProperties.color = enabled ? self.view.tintColor : UIColor.secondaryLabelColor;
    cell.contentConfiguration = configuration;
    cell.selectionStyle = enabled ? UITableViewCellSelectionStyleDefault : UITableViewCellSelectionStyleNone;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (indexPath.section >= SetupSectionCount) { return; }

    if (indexPath.section == SetupSectionCreateSession && indexPath.row == 1 && !self.flowController.isBusy) {
        [self.flowController createSession];
    } else if (indexPath.section == SetupSectionRetrieveSession && indexPath.row == 1
               && !self.flowController.isBusy && self.pin.length > 0) {
        [self.flowController retrieveSessionWithPIN:self.pin];
    }
}

- (void)contactEmailChanged:(UITextField *)field {
    self.flowController.session.contactEmail = field.text ?: @"";
}

- (void)pinChanged:(UITextField *)field {
    self.pin = field.text ?: @"";
    [self.tableView reloadSections:[NSIndexSet indexSetWithIndex:SetupSectionRetrieveSession]
                  withRowAnimation:UITableViewRowAnimationNone];
}

- (void)bindFlowController {
    __weak typeof(self) weakSelf = self;
    self.flowController.onBusyChanged = ^(BOOL busy) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        (void)busy;
        [strongSelf.tableView reloadData];
    };
    self.flowController.onErrorChanged = ^(NSString * _Nullable errorMessage) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        (void)errorMessage;
        [strongSelf.tableView reloadData];
    };
}

@end

File: SamplePresence/UIKit/SetupViewController.swift
import UIKit

final class SetupViewController: UITableViewController {
    private let flowController: DemoFlowController
    private var pin = ""

    private enum Section: Int, CaseIterable {
        case authToken
        case createSession
        case retrieveSession
    }

    init(flowController: DemoFlowController) {
        self.flowController = flowController
        super.init(style: .insetGrouped)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Setup Session"
        tableView.register(FormFieldCell.self, forCellReuseIdentifier: FormFieldCell.reuseIdentifier)
        bindFlowController()
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        flowController.errorMessage == nil ? Section.allCases.count : Section.allCases.count + 1
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch Section(rawValue: section) {
        case .authToken: return "Auth Token"
        case .createSession: return "Create Session"
        case .retrieveSession: return "Retrieve Session"
        default: return nil
        }
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == Section.allCases.count {
            return 1
        }
        switch Section(rawValue: section) {
        case .authToken: return 1
        case .createSession: return 2
        case .retrieveSession: return 2
        default: return 0
        }
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == Section.allCases.count {
            let cell = UITableViewCell(style: .default, reuseIdentifier: "ErrorCell")
            cell.selectionStyle = .none
            cell.textLabel?.text = flowController.errorMessage
            cell.textLabel?.textColor = .systemRed
            cell.textLabel?.numberOfLines = 0
            return cell
        }

        switch Section(rawValue: indexPath.section) {
        case .authToken:
            let cell = UITableViewCell(style: .default, reuseIdentifier: "TokenCell")
            cell.selectionStyle = .none
            cell.textLabel?.font = .preferredFont(forTextStyle: .footnote)
            cell.textLabel?.numberOfLines = 0
            cell.textLabel?.text = flowController.session.authToken
            return cell
        case .createSession:
            if indexPath.row == 0 {
                guard let cell = tableView.dequeueReusableCell(withIdentifier: FormFieldCell.reuseIdentifier, for: indexPath) as? FormFieldCell else {
                    return UITableViewCell()
                }
                cell.configure(
                    title: "Contact Email:",
                    text: flowController.session.contactEmail,
                    placeholder: "contact@example.com",
                    keyboardType: .emailAddress
                )
                cell.valueField.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.contactEmail = cell.valueField.text ?? ""
                }, for: .editingChanged)
                return cell
            }
            return actionCell(title: flowController.isBusy ? "Creating…" : "Create Session", enabled: !flowController.isBusy)
        case .retrieveSession:
            if indexPath.row == 0 {
                guard let cell = tableView.dequeueReusableCell(withIdentifier: FormFieldCell.reuseIdentifier, for: indexPath) as? FormFieldCell else {
                    return UITableViewCell()
                }
                cell.configure(title: "PIN:", text: pin, placeholder: "Required")
                cell.valueField.addAction(UIAction { [weak self] action in
                    guard let field = action.sender as? UITextField else { return }
                    self?.pin = field.text ?? ""
                }, for: .editingChanged)
                return cell
            }
            return actionCell(
                title: flowController.isBusy ? "Retrieving…" : "Retrieve Session",
                enabled: !flowController.isBusy && !pin.isEmpty
            )
        default:
            return UITableViewCell()
        }
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        guard indexPath.section < Section.allCases.count else { return }

        switch Section(rawValue: indexPath.section) {
        case .createSession where indexPath.row == 1:
            guard !flowController.isBusy else { return }
            Task { await flowController.createSession() }
        case .retrieveSession where indexPath.row == 1:
            guard !flowController.isBusy, !pin.isEmpty else { return }
            Task { await flowController.retrieveSession(pin: pin) }
        default:
            break
        }
    }

    private func actionCell(title: String, enabled: Bool) -> UITableViewCell {
        let cell = UITableViewCell(style: .default, reuseIdentifier: nil)
        var configuration = UIListContentConfiguration.cell()
        configuration.text = title
        configuration.textProperties.alignment = .center
        configuration.textProperties.color = enabled ? view.tintColor : .secondaryLabel
        cell.contentConfiguration = configuration
        cell.selectionStyle = enabled ? .default : .none
        return cell
    }

    private func bindFlowController() {
        flowController.onBusyChanged = { [weak self] _ in
            self?.tableView.reloadData()
        }
        flowController.onErrorChanged = { [weak self] _ in
            self?.tableView.reloadData()
        }
    }
}
